metric_learn/_util.py

```diff
@@ -411,3 +426,298 @@ def validate_vector(u, dtype=None):
+                       'linear transformation `init` must match the '
+                       'dimensionality of the given inputs `X` ({}).'
+                       .format(init.shape[1], n_features))
+
+    # Assert that init.shape[0] <= init.shape[1]
+    if init.shape[0] > init.shape[1]:
+      raise ValueError('The output dimensionality ({}) of the given '
+                       'linear transformation `init` cannot be '
+                       'greater than its input dimensionality ({}).'
+                       .format(init.shape[0], init.shape[1]))
+
+    # Assert that self.num_dims = init.shape[0]
+    if num_dims != init.shape[0]:
+      raise ValueError('The preferred dimensionality of the '
+                       'projected space `num_dims` ({}) does'
+                       ' not match the output dimensionality of '
+                       'the given linear transformation '
+                       '`init` ({})!'
+                       .format(num_dims,
+                               init.shape[0]))
+  elif init in authorized_inits:
+    pass
+  else:
+    raise ValueError(
+        "`init` must be '{}' "
+        "or a numpy array of shape (num_dims, n_features)."
+        .format("', '".join(authorized_inits)))
+
+  random_state = check_random_state(random_state)
+  transformation = init
+  if isinstance(init, np.ndarray):
+    pass
+  else:
+    n_samples = input.shape[0]
+    if init == 'auto':
+      if has_classes:
+        n_classes = len(np.unique(y))
+      if (has_classes and num_dims <= min(n_features, n_classes - 1)):
+        init = 'lda'
+      elif num_dims < min(n_features, n_samples):
+        init = 'pca'
+      else:
+        init = 'identity'
+    if init == 'identity':
+      transformation = np.eye(num_dims, input.shape[-1])
+    elif init == 'random':
+      transformation = random_state.randn(num_dims,
+                                          input.shape[-1])
+    elif init in {'pca', 'lda'}:
+      init_time = time.time()
+      if init == 'pca':
+        pca = PCA(n_components=num_dims,
+                  random_state=random_state)
+        if verbose:
+          print('Finding principal components... ')
+          sys.stdout.flush()
+        pca.fit(input)
+        transformation = pca.components_
+      elif init == 'lda':
+        lda = LinearDiscriminantAnalysis(n_components=num_dims)
+        if verbose:
+          print('Finding most discriminative components... ')
+          sys.stdout.flush()
+        lda.fit(input, y)
+        transformation = lda.scalings_.T[:num_dims]
+      if verbose:
+        print('done in {:5.2f}s'.format(time.time() - init_time))
+  return transformation
+
+
+def _initialize_metric_mahalanobis(input, init='identity', random_state=None,
+                                   return_inverse=False, strict_pd=False,
+                                   matrix_name='matrix'):
+  """Returns a standard mahalanobis matrix that can be used as a prior or an

```

Comment:
a standard mahalanobis --> a PSD
and add "as a prior or an initialization for the Mahalanobis distance"

Reply:
thanks, done